### PR TITLE
#623 Add support of window functions in Exposed DSL

### PR DESCRIPTION
**Pull request by Legohuman**
-Support of partition by and order by clauses
-Support of window frame clause (without EXCLUDE)
-Factories for common window functions
-Support for using aggregate functions as window functions

### CODE REVIEW

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/WindowFunction.kt [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+     */
+    fun groups(
+        start: CurrentOrPreceding,
+    ): WindowFunctionDefinition<T> = apply {
+        frameClause = WindowFrameClause(WindowFrameUnit.GROUPS, start, null)
+    }
+
+    override fun toQueryBuilder(queryBuilder: QueryBuilder) = queryBuilder {
+        function.toQueryBuilder(this)
+        +" OVER("
```
Comment (AlexeySoshin) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/WindowFunction.kt:119]: Thank you for your submission. Could you please run a linter on this code, though?

Comment (Legohuman) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/WindowFunction.kt:119]: I ran detekt linter and fixed warnings in new code. Although it sees no problem on that particular line.

**Comment by AlexeySoshin** [comment]
@Tapac What are your thoughts on that? 😃

**Review by AlexeySoshin — APPROVED** [review]

**Review by joc-a — COMMENTED** [review]
Thank you for your contribution! I left some comments.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/WindowFunction.kt [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+ */
+class NthValue<T>(
+    /** Returns the expression to evaluate. */
+    val expr: ExpressionWithColumnType<T>,
+    /** Returns the row index to find. */
+    val index: ExpressionWithColumnType<Int>
+) : WindowFunction<T?> {
+    override fun toQueryBuilder(queryBuilder: QueryBuilder) = queryBuilder {
+        append("NTH_VALUE(", expr, ", ", index, ")")
+    }
```
[the remark this answers is not in the document]
Comment (Legohuman) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/WindowFunction.kt:453]: I think, you are right. 'index' seemed to be good name at first as it`s database neutral, but the most of supported DBs define this argument as 'n' or 'num_row'. So I accept your changes.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/SQLExpressionBuilder.kt [hunk 3]
```diff
@@ -90,6 +90,68 @@ fun <T : Any?> ExpressionWithColumnType<T>.varPop(scale: Int = 2): VarPop<T> = V
...
+ * Returns value evaluated at the row that is the last row of the window frame.
+ */
+fun <T> ExpressionWithColumnType<T>.lastValue(): LastValue<T> = LastValue(this)
+
+/**
+ * Returns value evaluated at the row that is the [index] row of the window frame
+ * (counting from 1); null if no such row
+ */
+fun <T> ExpressionWithColumnType<T>.nthValue(index: ExpressionWithColumnType<Int>): NthValue<T> = NthValue(this, index)
+
```
[the remark this answers is not in the document]
Comment (Legohuman) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/SQLExpressionBuilder.kt:154]: The same as https://github.com/JetBrains/Exposed/pull/1651#discussion_r1241028043

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/SQLExpressionBuilder.kt [hunk 4]
```diff
@@ -90,6 +90,68 @@ fun <T : Any?> ExpressionWithColumnType<T>.varPop(scale: Int = 2): VarPop<T> = V
...
+/**
+ * Returns value evaluated at the row that is the last row of the window frame.
+ */
+fun <T> ExpressionWithColumnType<T>.lastValue(): LastValue<T> = LastValue(this)
+
+/**
+ * Returns value evaluated at the row that is the [index] row of the window frame
+ * (counting from 1); null if no such row
+ */
+fun <T> ExpressionWithColumnType<T>.nthValue(index: ExpressionWithColumnType<Int>): NthValue<T> = NthValue(this, index)
```
[the remark this answers is not in the document]
Comment (Legohuman) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/SQLExpressionBuilder.kt]: I found that most databases support expressions in window function arguments. See org.jetbrains.exposed.sql.tests.shared.functions.WindowFunctionsTests#supportsExpressionsInWindowFunctionArguments

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/WindowFunction.kt [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+
+    fun toQueryBuilder(queryBuilder: QueryBuilder)
+}
+
+/** Represents an SQL window function frame bound that is CURRENT ROW or one of PRECEDING forms. */
+interface CurrentOrPreceding : WindowFrameBound
+
+/** Represents an SQL window function frame bound that is CURRENT ROW or one of FOLLOWING forms. */
+interface CurrentOrFollowing : WindowFrameBound
```
[the remark this answers is not in the document]
Comment (Legohuman) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/WindowFunction.kt:224]: CurrentOrPreceding is used in functions 
org.jetbrains.exposed.sql.WindowFunctionDefinition#rows(org.jetbrains.exposed.sql.CurrentOrPreceding)
org.jetbrains.exposed.sql.WindowFunctionDefinition#range(org.jetbrains.exposed.sql.CurrentOrPreceding)
org.jetbrains.exposed.sql.WindowFunctionDefinition#groups(org.jetbrains.exposed.sql.CurrentOrPreceding)
to limit posible argument 'start' to sane value. 
- 'start' can not be **UNBOUNDED FOLLOWING**
- 'end' can nor appear earlier in rows window than 'start'. Omitted 'end' defaults to **CURRENT ROW**, so using **offset FOLLOWING** for 'start' in these functions is meaningless.
CurrentOrPreceding interface reflects these limitations. They are same for all databases, though can be implemented slightly different: query syntax validation or execution with empty result.
CurrentOrFollowing interface currently is not used in functions, it is rather for WindowFrameBound hierarchy completeness.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/SQLExpressionBuilder.kt [hunk 6]
```diff
@@ -90,6 +90,68 @@ fun <T : Any?> ExpressionWithColumnType<T>.varPop(scale: Int = 2): VarPop<T> = V
  */
 fun <T : Any?> ExpressionWithColumnType<T>.varSamp(scale: Int = 2): VarSamp<T> = VarSamp(this, scale)
 
+// Window Functions
+
+/** Returns the number of the current row within its partition, counting from 1. */
+fun rowNumber(): RowNumber = RowNumber()
```
Comment (e5l) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/SQLExpressionBuilder.kt:96]: Please avoid adding top-level functions for API-specific stuff. They will appear in code completion for every application with Exposed dependency

**Review by e5l — CHANGES_REQUESTED** [review]

**Comment by e5l** [comment]
Could you also rebase changes on the latest `main`?

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/SQLExpressionBuilder.kt (hunk 6 above)
Comment (Legohuman) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/SQLExpressionBuilder.kt]: In last commit I moved new window functions to ISqlExpressionBuilder. I should mention that some existing functions are also defined as top-level though and some are in ISqlExpressionBuilder. I didn`t grasp initial intention of this separation.

**Comment by e5l** [comment]
@bog-walk, could you take a look?

File: exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/functions/WindowFunctionsTests.kt [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+    private val supportsNthValueFunction = values().toList() - listOf(SQLSERVER)
+    private val supportsExpressionsInWindowFunctionArguments = values().toList() - listOf(MYSQL)
+    private val supportsExpressionsInWindowFrameClause = values().toList() - listOf(MYSQL, SQLSERVER, MARIADB)
+    private val supportsDefaultValueInLeadLagFunctions = values().toList() - listOf(MARIADB)
+    private val supportsRangeModeWithOffsetFrameBound = values().toList() - listOf(SQLSERVER)
+
+    @Suppress("LongMethod")
+    @Test
+    fun testWindowFunctions() = withSales { testDb, sales ->
+        if (currentDialect.supportsWindowFunctions) {
```
Comment (bog-walk) [inline exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/functions/WindowFunctionsTests.kt:41]: Is MySQL (versions earlier than 8.0) the only Exposed database not supporting window functions?

If so, please replace these `currentDialect` checks with the more descriptive conditional from DatabaseTestsBase.kt, `isOldMySql()`. That way it becomes immediately obvious which databases are being included/excluded in each test without having to leave the test suite and/or search the individual vendor files for `supportsWindowFunctions` overrides.

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/WindowFunction.kt [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+     * are processed by the window function.
+     */
+    fun orderBy(column: Expression<*>, order: SortOrder = SortOrder.ASC): WindowFunctionDefinition<T> =
+        orderBy(column to order)
+
+    /**
+     * Defines sorting order by column and order pairs [order] in which the rows of a partition
+     * are processed by the window function.
+     */
+    fun orderBy(vararg order: Pair<Expression<*>, SortOrder>): WindowFunctionDefinition<T> = apply {
```
[the remark this answers is not in the document]
Comment (Legohuman) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/WindowFunction.kt:50]: Yes, you are absolutely correct. 
Several invocations of `partitionBy()` / `orderBy()` will augment respective lists.
Each invocation of `rows()` / `range()` / `groups()`will redefine `frameClause`.
Last code snippet will use only `range()` clause.
I implemented API in such relaxed style similarly to `Query.groupBy()` / `AbstractQuery.orderBy()` (list augmentation) and `AbstractQuery.limit()` (value redefinition)

File: exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/functions/WindowFunctionsTests.kt (hunk 7 above)
Comment (Legohuman) [reply exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/functions/WindowFunctionsTests.kt]: It is good point, better to use this recently added helper function. I will refactor these conditions.

**Review by e5l — APPROVED** [review]

**Pull request merged**